## Анализ данных об объектах культурного наследия Ленинградской области

**Что мы будем делать:**
1. Загрузим и исследуем датасет с объектами культурного наследия.
2. Очистим данные от пропусков и "мусора".
3. Проведем разведочный анализ данных (EDA).
4. Научимся извлекать полезную информацию из сложных столбцов.

#### Часть 1. Подготовка окружения (5 минут)

Импортируем необходимые библиотеки.

In [2]:
import pandas as pd
import ast

# Настройка отображения pandas (показываем все столбцы)
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 20)

#### Часть 2. Загрузка и первичный осмотр данных (10 минут)

Загрузим данные из JSON файла.

In [3]:
# Загружаем данные
okn = pd.read_json('okn.json')

# Смотрим на первые 5 строк
okn.head()

,fid,Объект,Номер в реестре,название улицы,Комментарий,Идентификатор дома в ФИАС,FiasAuxSubobjId,FiasAuxObjId,Идентификатор улицы в ФИАС,Идентификатор населенного пункта в ФИАС,Идентификатор внутрегородской территории в ФИАС,Идентификатор города в ФИАС,Идентификатор городских и сельских поселений в ФИАС,Идентификатор района в ФИАС,Идентификатор автономного округа в ФИАС,Идентификатор региона в ФИАС,Идентификатор страны в ФИАС,Полный адрес,Адрес постановки,Id-Регион,Регион,учетный номер,Id-Категория историко-культурного значения,Категория историко-культурного значения,Id-Тип объекта культурного наследия,Тип объекта культурного наследия,Id-Вид объекта,Вид объекта,описание предмета охраны,текстовое описание границ,Id-общая видовая принадлежность,Id-Принадлежность к Юнеско,Принадлежность к Юнеско,Id-Особо ценный объект,Особо ценный объект,ВОВ-id,Принадлежность к ВОВ,идентификатор ансамбля,дата создания,GeoId,X,Y,наименование документа,зоны охраны и режимы использования земель,Items,Id - идентификатор,признак актуальности,признак публичности,год постановки ОКН на Госохрану,год проведения реставрации,"параметры, идентифицирующие ФЦП",относительный URL объекта,категория объекта,дата создания объекта,дата обновления объекта,На_карте,Изображение_str,муницип_zona_oxr,муницип_zouit,фед_zouit,фед_zona_oxran,адрес_из_документа
0,1,Жилой дом и комплекс приусадебных построек,471521428750005,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"Ленинградская область, Выборгский район, г. Вы...",NaN,47,Ленинградская область,47-57398,2,Регионального значения,NaN,NaN,2,Ансамбль,1. Градостроительные характеристикиОбъемно­ пр...,NaN,"[ { ""id"": ""3"", ""value"": ""Памятник градостроите...",2,нет,2,нет,False,Нет,NaN,"начало XIX в., 1901 г.",NaN,28.735884,60.712701,"[ { ""archive"": { ""id"": ""815411"", ""url"": ""https...",NaN,NaN,200339,True,1.0,NaN,NaN,NaN,/cdm/v2/heritages/1/200339,heritage,1432252800000,1432252800000,"{ ""coordinates"": [ 28.735883999999999, 60.7127...","{ ""title"": ""5542721"", ""url"": ""https:\/\/okn-mk...",None,None,None,None,
1,2,"Дом, где в сентябре – октябре 1917 г. жил и ра...",471511428740005,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"Ленинградская область, Выборгский муниципальны...",NaN,47,Ленинградская область,47-58157,2,Регионального значения,NaN,NaN,1,Памятник,Предметом охраны объекта культурного наследия ...,NaN,"[ { ""id"": ""2"", ""value"": ""Памятник истории"" } ]",2,нет,2,нет,False,Нет,NaN,дата создания не определена,NaN,28.7688589153427,60.698730887460016,"[ { ""archive"": { ""id"": ""846862"", ""url"": ""https...",NaN,NaN,200935,True,1.0,NaN,NaN,NaN,/cdm/v2/heritages/1/200935,heritage,1433721600000,1433721600000,"{ ""coordinates"": [ 28.768858915342701, 60.6987...","{ ""title"": ""5541600"", ""url"": ""https:\/\/okn-mk...",None,None,None,None,"Ленинградская область, г. Выборг, Рубежная ул...."
2,3,"Дом доходный, рубеж XIX-XX вв. Здесь в сентябр...",471511428730005,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"Ленинградская область, Выборгский муниципальны...",NaN,47,Ленинградская область,47-58170,2,Регионального значения,NaN,NaN,1,Памятник,Градостроительные характеристики:Объемно­ прос...,NaN,"[ { ""id"": ""2"", ""value"": ""Памятник истории"" } ]",2,нет,2,нет,False,Нет,NaN,сентябрь 1917 г.,NaN,28.76056755952353,60.70998294066887,"[ { ""archive"": { ""id"": ""5537709"", ""url"": ""http...",NaN,NaN,200948,True,1.0,NaN,NaN,NaN,/cdm/v2/heritages/1/200948,heritage,1433721600000,1433721600000,"{ ""coordinates"": [ 28.76056755952353, 60.70998...","{ ""title"": ""847033"", ""url"": ""https:\/\/okn-mk....",None,None,None,None,
3,4,Школа промышленная и лесотехническая,471511430820005,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"Ленинградская область, Выборгский муниципальны...",NaN,47,Ленинградская область,47-60540,2,Регионального значения,NaN,NaN,1,Памятник,Предметом охраны объекта культурного наследия ...,NaN,"[ { ""id"": ""3"", ""value"": ""Памятник градостроите...",2,не

#### Задание 1 

**Ответьте на вопросы:**
1. Сколько строк и столбцов в датасете?
2. Какие типы данных встречаются?
3. Есть ли пропуски (null) в данных?

**Подсказка:** Используйте методы `.info()` и `.shape`.

In [4]:
print(f"Форма датасета: {okn.shape}")
print(f"Строк: {okn.shape[0]}, Столбцов: {okn.shape[1]}")
okn.info()

Форма датасета: (2615, 62)
Строк: 2615, Столбцов: 62
<class 'pandas.core.frame.DataFrame'>
Index: 2615 entries, 0 to 2614
Data columns (total 62 columns):
 #   Column                                               Non-Null Count  Dtype  
---  ------                                               --------------  -----  
 0   fid                                                  2615 non-null   int64  
 1   Объект                                               2615 non-null   object 
 2   Номер в реестре                                      2615 non-null   object 
 3   название улицы                                       0 non-null      float64
 4   Комментарий                                          0 non-null      float64
 5   Идентификатор дома в ФИАС                            0 non-null      float64
 6   FiasAuxSubobjId                                      0 non-null      float64
 7   FiasAuxObjId                                         0 non-null      float64
 8   Идентификатор улицы 

#### Часть 3. Очистка данных

Реальные данные редко бывают идеальными. Нужно:
- Удалить столбцы, полностью состоящие из пропусков.
- Создать копию DataFrame, чтобы не испортить исходные данные.

In [5]:
# Создаем копию для работы
okn_cleaned = okn.copy()

# Удаляем столбцы, где все значения NaN (how='all')
okn_cleaned = okn_cleaned.dropna(axis=1, how='all')

print(f"Было столбцов: {okn.shape[1]}, стало: {okn_cleaned.shape[1]}")
okn_cleaned.info()

Было столбцов: 62, стало: 37
<class 'pandas.core.frame.DataFrame'>
Index: 2615 entries, 0 to 2614
Data columns (total 37 columns):
 #   Column                                      Non-Null Count  Dtype  
---  ------                                      --------------  -----  
 0   fid                                         2615 non-null   int64  
 1   Объект                                      2615 non-null   object 
 2   Номер в реестре                             2615 non-null   object 
 3   Полный адрес                                2239 non-null   object 
 4   Id-Регион                                   2615 non-null   int64  
 5   Регион                                      2615 non-null   object 
 6   учетный номер                               2615 non-null   object 
 7   Id-Категория историко-культурного значения  2615 non-null   int64  
 8   Категория историко-культурного значения     2615 non-null   object 
 9   Id-Вид объекта                              2615 non-null   i

#### Задание 2 

Посмотрите на столбцы `'На_карте'` и `'Изображение_str'`. Что в них хранится? Какой у них тип данных?

In [6]:
print("Столбец 'На_карте':")
print(okn_cleaned['На_карте'].head(3))
print(f"Тип данных: {okn_cleaned['На_карте'].dtype}")

Столбец 'На_карте':
0    { "coordinates": [ 28.735883999999999, 60.7127...
1    { "coordinates": [ 28.768858915342701, 60.6987...
2    { "coordinates": [ 28.76056755952353, 60.70998...
Name: На_карте, dtype: object
Тип данных: object


In [7]:
print("\nСтолбец 'Изображение_str':")
print(okn_cleaned['Изображение_str'].head(3))
print(f"Тип данных: {okn_cleaned['Изображение_str'].dtype}")


Столбец 'Изображение_str':
0    { "title": "5542721", "url": "https:\/\/okn-mk...
1    { "title": "5541600", "url": "https:\/\/okn-mk...
2    { "title": "847033", "url": "https:\/\/okn-mk....
Name: Изображение_str, dtype: object
Тип данных: object


#### Часть 4. Работа со сложными столбцами

Некоторые столбцы хранят словари в виде строк. Например, `'Id-общая видовая принадлежность'` содержит JSON-подобные строки. Нужно извлечь оттуда нужное значение `'value'`.

In [8]:
# Посмотрим, что в столбце 'Id-общая видовая принадлежность'
okn_cleaned['Id-общая видовая принадлежность'].head(10)

0    [ { "id": "3", "value": "Памятник градостроите...
1       [ { "id": "2", "value": "Памятник истории" } ]
2       [ { "id": "2", "value": "Памятник истории" } ]
3    [ { "id": "3", "value": "Памятник градостроите...
4    [ { "id": "1", "value": "Памятник археологии" } ]
5    [ { "id": "1", "value": "Памятник археологии" } ]
6    [ { "id": "1", "value": "Памятник археологии" } ]
7    [ { "id": "1", "value": "Памятник археологии" } ]
8    [ { "id": "1", "value": "Памятник археологии" } ]
9    [ { "id": "1", "value": "Памятник археологии" } ]
Name: Id-общая видовая принадлежность, dtype: object

#### Задание 3

Напишите функцию `extract_value()`, которая:
- Принимает значение из столбца.
- Если значение — NaN, возвращает None.
- Если значение — строка, похожая на словарь (содержит `{` и `}`), преобразует её в словарь с помощью `ast.literal_eval()` и возвращает значение по ключу `'value'`.
- Если строка — список словарей (начинается с `[`), возьмите первый словарь и извлеките `'value'`.
- В случае ошибки возвращает None.

Затем примените эту функцию к столбцу `'Id-общая видовая принадлежность'` и сохраните результат в новый столбец `'Видовая принадлежность'`.

In [9]:
def extract_value(value):
    if pd.isna(value):
        return None
    
    try:
        value_str = str(value).strip()
        if value_str.startswith('[') and value_str.endswith(']'):
            dict_list = value_str[1:-1]
        else:
            dict_list = value_str
        
        if not dict_list:
            return None
        
        dictionary = ast.literal_eval(dict_list)
        
        if isinstance(dictionary, dict):
            return dictionary.get('value')
        elif isinstance(dictionary, tuple):
            return 'Памятник двойной принадлежности'
        else:
            return None
            
    except (ValueError, SyntaxError, TypeError, AttributeError) as e:
        print(f"Ошибка: {e} в значении: {value}")
        return None

In [18]:
okn_cleaned['Видовая принадлежность'] = okn_cleaned['Id-общая видовая принадлежность'].map(extract_value)
okn_cleaned['Видовая принадлежность']

0       Памятник градостроительства и архитектуры
1                                Памятник истории
2                                Памятник истории
3       Памятник градостроительства и архитектуры
4                             Памятник археологии
                          ...                    
2610                             Памятник истории
2611                             Памятник истории
2612                             Памятник истории
2613    Памятник градостроительства и архитектуры
2614                             Памятник истории
Name: Видовая принадлежность, Length: 2615, dtype: object

#### Часть 5. Разведочный анализ данных (EDA)

Теперь, когда данные очищены, ответим на несколько вопросов о наборе данных.

#### Задание 4

**Вопрос:** Сколько объектов относятся к Великой Отечественной войне (столбец `'Принадлежность к ВОВ'`)?

Подсказка: используйте `.value_counts()`.

In [21]:
vov_counts = okn_cleaned['Принадлежность к ВОВ'].value_counts()
print("Распределение по принадлежности к ВОВ:")
print(vov_counts)
print(f"\nОбъектов, связанных с ВОВ: {vov_counts['Да']}")

Распределение по принадлежности к ВОВ:
Принадлежность к ВОВ
Нет    1833
Да      782
Name: count, dtype: int64

Объектов, связанных с ВОВ: 782


#### Задание 5

**Вопрос:** Какие категории историко-культурного значения встречаются? В каком количестве?

Используйте столбец `'Категория историко-культурного значения'`.

In [22]:
category_counts = okn_cleaned['Категория историко-культурного значения'].value_counts()
print("Распределение по категориям:")
print(category_counts)

Распределение по категориям:
Категория историко-культурного значения
Регионального значения                1830
Федерального значения                  722
Местного (муниципального) значения      63
Name: count, dtype: int64


#### Задание 6

**Вопрос:** Как распределены объекты по видовой принадлежности? Посчитайте процентное соотношение.

Подсказка: используйте `value_counts(normalize=True)`.

In [23]:
type_counts = okn_cleaned['Видовая принадлежность'].value_counts()
type_percents = okn_cleaned['Видовая принадлежность'].value_counts(normalize=True) * 100
print("Распределение по видовой принадлежности:")
print(type_counts)
print("\nПроцентное соотношение:")
print(type_percents.round(2))

Распределение по видовой принадлежности:
Видовая принадлежность
Памятник истории                             1235
Памятник градостроительства и архитектуры     844
Памятник археологии                           319
Памятник религии                              145
Памятник монументального искусства             19
Памятник двойной принадлежности                16
Name: count, dtype: int64

Процентное соотношение:
Видовая принадлежность
Памятник истории                             47.91
Памятник градостроительства и архитектуры    32.74
Памятник археологии                          12.37
Памятник религии                              5.62
Памятник монументального искусства            0.74
Памятник двойной принадлежности               0.62
Name: proportion, dtype: float64


#### Задание 7

**Вопрос:** Какие виды объектов (`'Вид объекта'`) встречаются чаще всего? Выведите топ-5.

In [27]:
top5_types = okn_cleaned['Вид объекта'].value_counts()
print(top5_types)

Вид объекта
Памятник                     2141
Ансамбль                      417
Достопримечательное место      57
Name: count, dtype: int64


#### Часть 6. Фильтрация и поиск

#### Задание 8

**Вопрос:** Найдите объекты, которые одновременно относятся к наследию ЮНЕСКО (`'Принадлежность к Юнеско' == 'да'`) и к Великой Отечественной войне (`'Принадлежность к ВОВ' == 'Да'`).

Выведите их названия (столбец `'Объект'`).

In [29]:
mask = (okn_cleaned['Принадлежность к Юнеско'] == 'да') & (okn_cleaned['Принадлежность к ВОВ'] == 'Да')
result = okn_cleaned[mask]
print(f"Найдено объектов: {len(result)}")
print("\nНазвания объектов:")
result['Объект'].tolist()

Найдено объектов: 27

Названия объектов:


['Памятные километровые столбы на «Дороге жизни»',
 'Памятник «Румболовская гора» на «Дороге жизни»',
 'Памятник «Сад мира»',
 'Памятник «Безымянная высота»',
 'Памятник «Лемболовская твердыня»',
 'Памятник «Разорванное кольцо» на «Дороге жизни»',
 'Памятник «Катюша» на «Дороге жизни»',
 'Памятник-стела воинам-понтонерам, обеспечивавшим транспортную связь осажденного Ленинграда со страной в 1943 году',
 'Памятник железнодорожникам, обеспечивавшим перевозку грузов для осажденного Ленинграда по «Дороге жизни»',
 'Памятник ленинградским детям «Цветок жизни» на «Дороге жизни»',
 'Памятный знак – «Автомашина ЗИС-5», установленный на «Дороге жизни», которая проходила здесь в 1941-1942 гг.',
 'Памятник «Берег мужественных» на «Ораниенбаумском пятачке»',
 'Ансамбль «Зеленый пояс Славы Ленинграда»',
 'Памятник «Ивановский пятачок»',
 'Памятник «Невский порог»',
 'Памятник-стела на «Дороге жизни»',
 'Памятник «Невский пятачок»',
 'Памятник «Прорыв»',
 'Памятник «Штурм»',
 'Памятник «Атака» на «О

#### Задание 9

**Вопрос:** Найдите все объекты, в названии которых есть слово "Дорога жизни".

Подсказка: используйте `.str.contains()`.

In [33]:
mask = okn_cleaned['Объект'].str.contains('Дорог', case=True, na=False)
result = okn_cleaned[mask]
print(f"Найдено объектов с упоминанием 'Дорога жизни': {len(result)}")
print("\nНазвания объектов:")
for name in result['Объект'].tolist():
    print(f"  - {name}")

Найдено объектов с упоминанием 'Дорога жизни': 19

Названия объектов:
  - Достопримечательное место "Дорога Жизни"
  - Памятный знак-стела, посвящен «Дороге жизни», которая проходила здесь в 1941-43 гг.
  - Памятные километровые столбы на «Дороге жизни»
  - Памятник «Румболовская гора» на «Дороге жизни»
  - Памятник «Разорванное кольцо» на «Дороге жизни»
  - Памятник «Катюша» на «Дороге жизни»
  - Памятник железнодорожникам, обеспечивавшим перевозку грузов для осажденного Ленинграда по «Дороге жизни»
  - Памятник ленинградским детям «Цветок жизни» на «Дороге жизни»
  - Памятный знак – «Автомашина ЗИС-5», установленный на «Дороге жизни», которая проходила здесь в 1941-1942 гг.
  - Памятник-стела, посвященный «Дороге жизни», которая проходила здесь в 1941 г. в период боев за г. Тихвин. В Заборье в то время находился важный автомобильно-железнодорожный перевалочный пункт
  - Памятник-стела на «Дороге жизни»
  - Мемориал героям Ладоги, посвящен морякам Ладожской военной флотилии и речникам

#### Задание 10

**Вопрос:** Сколько объектов не имеют координат (столбец `'X'` или `'Y'` пуст)?

Подсказка: используйте `.isna()`.

In [34]:
no_coords = okn_cleaned[okn_cleaned['X'].isna() | okn_cleaned['Y'].isna()]
print(f"Объектов без координат: {len(no_coords)}")
print(f"Всего объектов: {len(okn_cleaned)}")
print(f"Процент: {len(no_coords) / len(okn_cleaned) * 100:.1f}%")

Объектов без координат: 351
Всего объектов: 2615
Процент: 13.4%


**Ключевые выводы по датасету**:

- В датасете **2615 объектов** культурного наследия Ленинградской области
- Большинство объектов имеют **региональное значение** (около 70%)
- Около **30% объектов** связаны с Великой Отечественной войной
- Самые распространённые виды объектов: **памятники** и **ансамбли**
- Основная видовая принадлежность: **памятники истории** (около 48%)